handler/oauth2/strategy_jwt.go

```diff
@@ -36,66 +38,99 @@ func (h RS256JWTStrategy) AuthorizeCodeSignature(token string) string {
 	return h.signature(token)
 }
 
+func (h *RS256JWTStrategy) ValidateJWT(tokenType fosite.TokenType, token string) (requester fosite.Requester, err error) {
+	t, err := h.validate(token)
+	if err != nil {
+		return nil, err
+	}
+
+	claims := jwt.JWTClaims{}
+	claims.FromMapClaims(t.Claims.(jwtx.MapClaims))
+
+	requester = &fosite.Request{
+		Client:      &fosite.DefaultClient{},
+		RequestedAt: claims.IssuedAt,
+		Session: &JWTSession{
+			JWTClaims: &claims,
+			JWTHeader: &jwt.Headers{
+				Extra: make(map[string]interface{}),
+			},
+			ExpiresAt: map[fosite.TokenType]time.Time{
+				tokenType: claims.ExpiresAt,
+			},
+			Subject: claims.Subject,
+		},
+		Scopes:        claims.Scope,
+		GrantedScopes: claims.Scope,
+	}
+
+	return
+}
+
 func (h *RS256JWTStrategy) GenerateAccessToken(_ context.Context, requester fosite.Requester) (token string, signature string, err error) {
 	return h.generate(fosite.AccessToken, requester)
 }
 
 func (h *RS256JWTStrategy) ValidateAccessToken(_ context.Context, _ fosite.Requester, token string) error {
-	return h.validate(token)
+	_, err := h.validate(token)
+	return err
 }
 
 func (h *RS256JWTStrategy) GenerateRefreshToken(_ context.Context, requester fosite.Requester) (token string, signature string, err error) {
 	return h.generate(fosite.RefreshToken, requester)
 }
 
 func (h *RS256JWTStrategy) ValidateRefreshToken(_ context.Context, _ fosite.Requester, token string) error {
-	return h.validate(token)
+	_, err := h.validate(token)
+	return err
 }
 
 func (h *RS256JWTStrategy) GenerateAuthorizeCode(_ context.Context, requester fosite.Requester) (token string, signature string, err error) {
 	return h.generate(fosite.AuthorizeCode, requester)
 }
 
 func (h *RS256JWTStrategy) ValidateAuthorizeCode(_ context.Context, requester fosite.Requester, token string) error {
-	return h.validate(token)
+	_, err := h.validate(token)
+	return err
 }
 
-func (h *RS256JWTStrategy) validate(token string) error {
-	t, err := h.RS256JWTStrategy.Decode(token)
-	if err != nil {
-		return err
+func (h *RS256JWTStrategy) validate(token string) (t *jwtx.Token, err error) {
+	t, err = h.RS256JWTStrategy.Decode(token)
+
+	if err == nil {
+		err = t.Claims.Valid()
 	}
 
-	// validate the token
-	if err = t.Claims.Valid(); err != nil {
-		if e, ok := err.(*jwtx.ValidationError); ok {
+	if err != nil {
+		if e, ok := errors.Cause(err).(*jwtx.ValidationError); ok {
 			switch e.Errors {
 			case jwtx.ValidationErrorMalformed:
-				return errors.Wrap(fosite.ErrInvalidTokenFormat, err.Error())
+				err = errors.Wrap(fosite.ErrInvalidTokenFormat, err.Error())
```

Comment:
It definitely feels weird coming from other languages that just fall through, but Go breaks if you don't add a fallthrough statement: https://golang.org/ref/spec#Switch_statements